File: src/main/java/leetcode/editor/cn/PalindromeLinkedList.java
package leetcode.editor.cn;

import java.util.ArrayList;
import java.util.List;
import model.ListNode;

/**
 * [234] 回文链表
 */
public class PalindromeLinkedList {

	//leetcode submit region begin(Prohibit modification and deletion)

	/**
	 * Definition for singly-linked list. public class ListNode { int val; ListNode next; ListNode()
	 * {} ListNode(int val) { this.val = val; } ListNode(int val, ListNode next) { this.val = val;
	 * this.next = next; } }
	 */
	class Solution {

		public boolean isPalindrome(ListNode head) {
			List<ListNode> stack = new ArrayList<>();
			ListNode fast = head, slow = head;
			while (fast != null && fast.next != null) {
				stack.add(slow);
				fast = fast.next.next;
				slow = slow.next;
			}
			// 如果节点数为双数，中点不确定。 中点修正
			if (fast != null) {
				slow = slow.next;
			}
			while (slow != null) {
				ListNode topNode = stack.remove(stack.size() - 1);
				if (slow.val != topNode.val) {
					return false;
				}
				slow = slow.next;
			}

			return stack.isEmpty();
		}
	}
//leetcode submit region end(Prohibit modification and deletion)

}

File: src/main/java/leetcode/editor/cn/PalindromeLinkedList.go
  package leetcode.editor.cn;
  
  /**
   * [234] 回文链表
   */
  public class PalindromeLinkedList{
      public static void main(String[] args) {
           Solution solution = new PalindromeLinkedList().new Solution();
      }
      //leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
  // 1、快慢指针，找到中点
  fast, slow := head, head
  for fast != nil && fast.Next != nil {
    fast = fast.Next.Next
    slow = slow.Next
  }
  // 2、反转后半部分链表
  tail := reverse(slow)

  // 3、比较两个链表的值
  for head != nil && tail != nil {
    if head.Val != tail.Val {
      return false
    }
    head = head.Next
    tail = tail.Next
  }
  return true
}

func reverse(node *ListNode) *ListNode {
  if node == nil || node.Next == nil {
    return node
  }
  var p1 *ListNode
  p2, p3 := node, node.Next
  for p3 != nil {
    p2.Next = p1
    p1, p2, p3 = p2, p3, p3.Next
  }
  p2.Next = p1
  return p2
}
//leetcode submit region end(Prohibit modification and deletion)

  }